File: tools/semantic_ir/validation.py
"""STUNIR Semantic IR Validation.

DO-178C Level A Compliant
Validation functions with detailed error reporting.
"""

import re
from enum import Enum
from typing import Optional
from pydantic import BaseModel

from .ir_types import NodeID, IRHash, BinaryOperator, IRPrimitiveType
from .nodes import TypeReference, TypeKind, PrimitiveTypeRef
from .modules import IRModule


class ValidationStatus(str, Enum):
    """Validation status."""
    VALID = "valid"
    INVALID = "invalid"
    WARNING = "warning"


class ValidationResult(BaseModel):
    """Validation result with status and message."""
    status: ValidationStatus
    message: str = ""

    class Config:
        frozen = True

    @classmethod
    def valid(cls) -> "ValidationResult":
        """Create a valid result."""
        return cls(status=ValidationStatus.VALID)

    @classmethod
    def invalid(cls, message: str) -> "ValidationResult":
        """Create an invalid result with error message."""
        return cls(status=ValidationStatus.INVALID, message=message)

    @classmethod
    def warning(cls, message: str) -> "ValidationResult":
        """Create a warning result."""
        return cls(status=ValidationStatus.WARNING, message=message)


def validate_node_id(node_id: str) -> ValidationResult:
    """Validate node ID format.
    
    Node IDs must start with 'n_' followed by alphanumeric characters.
    """
    pattern = r"^n_[a-zA-Z0-9_]+$"
    if not re.match(pattern, node_id):
        return ValidationResult.invalid(
            f"Invalid node ID '{node_id}': must match pattern '{pattern}'"
        )
    return ValidationResult.valid()


def validate_hash(hash_str: str) -> ValidationResult:
    """Validate hash format.
    
    Hashes must be 'sha256:' followed by 64 hexadecimal characters.
    """
    pattern = r"^sha256:[a-f0-9]{64}$"
    if not re.match(pattern, hash_str):
        return ValidationResult.invalid(
            f"Invalid hash '{hash_str}': must match pattern '{pattern}'"
        )
    return ValidationResult.valid()


def validate_type_reference(type_ref: TypeReference) -> ValidationResult:
    """Validate type reference structure."""
    if type_ref.kind == TypeKind.PRIMITIVE:
        return ValidationResult.valid()
    elif type_ref.kind == TypeKind.REF:
        if not hasattr(type_ref, 'name') or not type_ref.name:
            return ValidationResult.invalid("Type reference must have non-empty name")
        return ValidationResult.valid()
    else:
        return ValidationResult.warning("Complex type validation not fully implemented")


def validate_module(module: IRModule) -> ValidationResult:
    """Validate IR module structure.
    
    Checks:
    - Valid node ID
    - Valid hash (if present)
    - Non-empty module name
    - Valid declarations
    """
    # Validate node ID
    id_result = validate_node_id(module.node_id)
    if id_result.status != ValidationStatus.VALID:
        return id_result

    # Validate hash if present
    if module.hash:
        hash_result = validate_hash(module.hash)
        if hash_result.status != ValidationStatus.VALID:
            return hash_result

    # Validate module name
    if not module.name:
        return ValidationResult.invalid("Module must have non-empty name")

    return ValidationResult.valid()


def check_type_compatibility(
    type1: TypeReference,
    type2: TypeReference
) -> bool:
    """Check if two types are compatible.
    
    Simplified type compatibility check.
    """
    if type1.kind != type2.kind:
        return False

    if isinstance(type1, PrimitiveTypeRef) and isinstance(type2, PrimitiveTypeRef):
        return type1.primitive == type2.primitive

    # For other types, basic kind matching
    return type1.kind == type2.kind


def check_binary_op_types(
    op: BinaryOperator,
    left: TypeReference,
    right: TypeReference
) -> ValidationResult:
    """Check type validity for binary operations.
    
    Ensures operand types are compatible with the operator.
    """
    # Both must be primitive types
    if not (isinstance(left, PrimitiveTypeRef) and isinstance(right, PrimitiveTypeRef)):
        return ValidationResult.invalid("Binary operators require primitive types")

    left_prim = left.primitive
    right_prim = right.primitive

    # Arithmetic operators
    if op in [BinaryOperator.ADD, BinaryOperator.SUB, BinaryOperator.MUL,
              BinaryOperator.DIV, BinaryOperator.MOD]:
        numeric_types = {
            IRPrimitiveType.I8, IRPrimitiveType.I16, IRPrimitiveType.I32, IRPrimitiveType.I64,
            IRPrimitiveType.U8, IRPrimitiveType.U16, IRPrimitiveType.U32, IRPrimitiveType.U64,
            IRPrimitiveType.F32, IRPrimitiveType.F64
        }
        if left_prim not in numeric_types or right_prim not in numeric_types:
            return ValidationResult.invalid("Arithmetic operators require numeric types")
        return ValidationResult.valid()

    # Comparison operators
    if op in [BinaryOperator.EQ, BinaryOperator.NEQ, BinaryOperator.LT,
              BinaryOperator.LEQ, BinaryOperator.GT, BinaryOperator.GEQ]:
        if not check_type_compatibility(left, right):
            return ValidationResult.invalid("Comparison operators require compatible types")
        return ValidationResult.valid()

    # Logical operators
    if op in [BinaryOperator.AND, BinaryOperator.OR]:
        if left_prim != IRPrimitiveType.BOOL or right_prim != IRPrimitiveType.BOOL:
            return ValidationResult.invalid("Logical operators require boolean types")
        return ValidationResult.valid()

    # Bitwise operators
    if op in [BinaryOperator.BIT_AND, BinaryOperator.BIT_OR, BinaryOperator.BIT_XOR,
              BinaryOperator.SHL, BinaryOperator.SHR]:
        integer_types = {
            IRPrimitiveType.I8, IRPrimitiveType.I16, IRPrimitiveType.I32, IRPrimitiveType.I64,
            IRPrimitiveType.U8, IRPrimitiveType.U16, IRPrimitiveType.U32, IRPrimitiveType.U64
        }
        if left_prim not in integer_types or right_prim not in integer_types:
            return ValidationResult.invalid("Bitwise operators require integer types")
        return ValidationResult.valid()

    # Assignment operator
    if op == BinaryOperator.ASSIGN:
        if not check_type_compatibility(left, right):
            return ValidationResult.invalid("Assignment requires compatible types")
        return ValidationResult.valid()

    return ValidationResult.warning(f"Unknown operator: {op}")

File: tools/rust/semantic_ir/src/validation.rs
//! STUNIR Semantic IR Validation
//!
//! Validation functions with detailed error reporting.

use serde::{Deserialize, Serialize};

use crate::types::*;
use crate::nodes::*;
use crate::modules::*;

/// Validation status
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ValidationStatus {
    Valid,
    Invalid,
    Warning,
}

/// Validation result
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ValidationResult {
    pub status: ValidationStatus,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub message: String,
}

impl ValidationResult {
    /// Create a valid result
    pub fn valid() -> Self {
        Self {
            status: ValidationStatus::Valid,
            message: String::new(),
        }
    }

    /// Create an invalid result
    pub fn invalid(message: impl Into<String>) -> Self {
        Self {
            status: ValidationStatus::Invalid,
            message: message.into(),
        }
    }

    /// Create a warning result
    pub fn warning(message: impl Into<String>) -> Self {
        Self {
            status: ValidationStatus::Warning,
            message: message.into(),
        }
    }

    /// Check if valid
    pub fn is_valid(&self) -> bool {
        self.status == ValidationStatus::Valid
    }
}

/// Validate node ID format
pub fn validate_node_id(node_id: &str) -> ValidationResult {
    if IRNodeBase::is_valid_node_id(node_id) {
        ValidationResult::valid()
    } else {
        ValidationResult::invalid(
            format!("Invalid node ID '{}': must start with 'n_'", node_id)
        )
    }
}

/// Validate hash format
pub fn validate_hash(hash: &str) -> ValidationResult {
    if IRNodeBase::is_valid_hash(hash) {
        ValidationResult::valid()
    } else {
        ValidationResult::invalid(
            format!("Invalid hash '{}': must be 'sha256:' + 64 hex chars", hash)
        )
    }
}

/// Validate type reference
pub fn validate_type_reference(type_ref: &TypeReference) -> ValidationResult {
    match type_ref {
        TypeReference::PrimitiveType { .. } => ValidationResult::valid(),
        TypeReference::TypeRef { name, .. } => {
            if name.is_empty() {
                ValidationResult::invalid("Type reference must have non-empty name")
            } else {
                ValidationResult::valid()
            }
        }
    }
}

/// Validate module structure
pub fn validate_module(module: &IRModule) -> ValidationResult {
    // Validate node ID
    let id_result = validate_node_id(&module.base.node_id);
    if !id_result.is_valid() {
        return id_result;
    }

    // Validate hash if present
    if let Some(ref hash) = module.base.hash {
        let hash_result = validate_hash(hash);
        if !hash_result.is_valid() {
            return hash_result;
        }
    }

    // Validate module name
    if module.name.is_empty() {
        return ValidationResult::invalid("Module must have non-empty name");
    }

    ValidationResult::valid()
}

/// Check type compatibility
pub fn check_type_compatibility(type1: &TypeReference, type2: &TypeReference) -> bool {
    match (type1, type2) {
        (
            TypeReference::PrimitiveType { primitive: p1 },
            TypeReference::PrimitiveType { primitive: p2 },
        ) => p1 == p2,
        (
            TypeReference::TypeRef { name: n1, .. },
            TypeReference::TypeRef { name: n2, .. },
        ) => n1 == n2,
        _ => false,
    }
}

/// Check binary operator type validity
pub fn check_binary_op_types(
    op: BinaryOperator,
    left: &TypeReference,
    right: &TypeReference,
) -> ValidationResult {
    // Both must be primitive types
    let (left_prim, right_prim) = match (left, right) {
        (
            TypeReference::PrimitiveType { primitive: l },
            TypeReference::PrimitiveType { primitive: r },
        ) => (l, r),
        _ => return ValidationResult::invalid("Binary operators require primitive types"),
    };

    use IRPrimitiveType::*;

    match op {
        BinaryOperator::Add
        | BinaryOperator::Sub
        | BinaryOperator::Mul
        | BinaryOperator::Div
        | BinaryOperator::Mod => {
            // Arithmetic operators need numeric types
            let numeric_types = [I8, I16, I32, I64, U8, U16, U32, U64, F32, F64];
            if numeric_types.contains(left_prim) && numeric_types.contains(right_prim) {
                ValidationResult::valid()
            } else {
                ValidationResult::invalid("Arithmetic operators require numeric types")
            }
        }
        BinaryOperator::Eq
        | BinaryOperator::Neq
        | BinaryOperator::Lt
        | BinaryOperator::Leq
        | BinaryOperator::Gt
        | BinaryOperator::Geq => {
            // Comparison operators need compatible types
            if check_type_compatibility(left, right) {
                ValidationResult::valid()
            } else {
                ValidationResult::invalid("Comparison operators require compatible types")
            }
        }
        BinaryOperator::And | BinaryOperator::Or => {
            // Logical operators need boolean types
            if *left_prim == Bool && *right_prim == Bool {
                ValidationResult::valid()
            } else {
                ValidationResult::invalid("Logical operators require boolean types")
            }
        }
        BinaryOperator::BitAnd
        | BinaryOperator::BitOr
        | BinaryOperator::BitXor
        | BinaryOperator::Shl
        | BinaryOperator::Shr => {
            // Bitwise operators need integer types
            let integer_types = [I8, I16, I32, I64, U8, U16, U32, U64];
            if integer_types.contains(left_prim) && integer_types.contains(right_prim) {
                ValidationResult::valid()
            } else {
                ValidationResult::invalid("Bitwise operators require integer types")
            }
        }
        BinaryOperator::Assign => {
            // Assignment requires compatible types
            if check_type_compatibility(left, right) {
                ValidationResult::valid()
            } else {
                ValidationResult::invalid("Assignment requires compatible types")
            }
        }
    }
}
